File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        bottom, top = 0, len(nums) - 1

        while bottom <= top:
            mid = (bottom + top) // 2

            if nums[mid] == target:
                return mid
            
            if nums[mid] > target:
                top = mid - 1
            else:
                bottom = mid + 1
        
        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-1.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {
        if (matrix.length === 0 || matrix[0].length === 0) {
            return false;
        }

        const rows = matrix.length, cols = matrix[0].length;
        let l = 0, r = rows * cols - 1;

        while (l <= r) {
            const mid = Math.floor((l + r) / 2);
            const row = Math.floor(mid / cols);
            const col = mid - row * cols;

            if (matrix[row][col] == target) {
                return true;
            }

            if (matrix[row][col] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return false;  
    }
}
